epoll/src/main/scala/fs2/io/epoll/unsafe/EpollSystem.scala

```diff
@@ -0,0 +1,255 @@
+      writeMutex.lock.surround {
+        def go(a: A, before: Int): IO[B] =
+          f(a).flatMap {
+            case Left(a) =>
+              IO(writeReadyCounter).flatMap { after =>
+                if (before != after)
+                  // there was a write-ready notification since we started, try again immediately
+                  go(a, after)
+                else
+                  IO.asyncCheckAttempt[Int] { cb =>
+                    IO {
+                      writeCallback = cb
+                      // check again before we suspend
+                      val now = writeReadyCounter
+                      if (now != before) {
+                        writeCallback = null
+                        Right(now)
+                      } else Left(Some(IO(this.writeCallback = null)))
+                    }
+                  }.flatMap(go(a, _))
+              }
+            case Right(b) => IO.pure(b)
+          }
+
+        IO(writeReadyCounter).flatMap(go(a, _))
+      }
+  }
+
+  final class Poller private[EpollSystem] (epfd: Int) {
+
+    private[this] val handles: HashMap[Long, PollHandle] = HashMap()
+    private[EpollSystem] def close(): Unit =
+      if (libc.jnr.close(epfd) != 0)
+        throw new IOException(strerror(errno()))
+
+    private[EpollSystem] def poll(timeout: Long): Boolean = {
+
+      val events = Memory.allocate(globalRuntime, EpollEvent.CStructSize * MaxEvents)
+      var polled = false
+
+      @tailrec
+      def processEvents(timeout: Int): Unit = {
+
+        val triggeredEvents = epoll_wait(epfd, events, MaxEvents, timeout)
+
+        if (triggeredEvents >= 0) {
+          polled = true
+
+          var i = 0
+          while (i < triggeredEvents) {
+            val epollEvent =
+              EpollEvent.apply(events, (i * EpollEvent.CStructSize).toLong, globalRuntime)
+            val handle = handles.get(epollEvent.data)
+
+            handle
+              .getOrElse(throw new IllegalStateException("Requested fd is not found"))
+              .notify(epollEvent.events.toInt)
+
+            i += 1
+          }
+        } else if (errno() != Errno.EINTR.intValue()) { // spurious wake-up by signal
+          throw new IOException(strerror(errno()))
+        }
+
+        if (triggeredEvents >= MaxEvents)
+          processEvents(0) // drain the ready list
+        else
+          ()
+      }
+
+      val timeoutMillis = if (timeout == -1) -1 else (timeout / 1000000).toInt
+      processEvents(timeoutMillis)
+
+      polled
+    }
+
+    private[EpollSystem] def needsPoll(): Boolean = !handles.isEmpty
+
+    private[EpollSystem] def register(
+        fd: Int,
+        reads: Boolean,
+        writes: Boolean,
+        handle: PollHandle,
+        cb: Either[Throwable, (PollHandle, IO[Unit])] => Unit
+    ): Unit = {
+      val events = (EPOLLET | (if (reads) EPOLLIN else 0) | (if (writes) EPOLLOUT else 0)).toInt
+      val data = (System.identityHashCode(handle).toLong << 32) | fd
```

Comment:
Hmm, not sure about this. `identityHashCode` is not guaranteed to be unique: two different handles may have the same one.

I think it is sufficient to use the `fd` to identify the handle. The expectation is that the user should not register the same `fd` more than once.